airflow-core/src/airflow/models/dagrun.py

```diff
@@ -1217,6 +1218,10 @@ def recalculate(self) -> _UnfinishedStates:
                     msg="success",
                 )
 
+            if (deadline := dag.deadline) and isinstance(deadline.reference, DeadlineReference.TYPES.DAGRUN):
+                # The dagrun has succeeded, so the deadline is no longer needed.
```

Comment:
```suggestion
                # The dagrun has succeeded, so the deadline(s) may no longer be needed.
```

Or just completely refactor.